src/westmere/ip4.h

```diff
@@ -0,0 +1,316 @@
+//
+// This function assumes that the processor supports SSE 4.1 instructions or better. That's true of most
+// processors in operation today (June 2023).
+//
+// See also sse_inet_aton_16 for a version that does not take a string length.
+static inline int sse_inet_aton(const char* ipv4_string, const size_t ipv4_string_length, uint8_t * destination) {
+  // This function always reads 16 bytes. With AVX-512 we can do a mask
+  // load, but it is not generally available with SSE 4.1.
+  const __m128i input = _mm_loadu_si128((const __m128i *)ipv4_string);
+  if (ipv4_string_length > 15) {
+    return 0;
+  }
+  // locate dots
+  uint16_t dotmask;
+  {
+    const __m128i dot = _mm_set1_epi8('.');
+    const __m128i t0 = _mm_cmpeq_epi8(input, dot);
+    dotmask = (uint16_t)_mm_movemask_epi8(t0);
+    uint16_t mask = (uint16_t)(1 << ipv4_string_length);
+    dotmask &= mask - 1;
+    dotmask |= mask;
+  }
+
+  // build a hashcode
+  const uint8_t hashcode = ((6639 * dotmask) >> 13);
+
+  // grab the index of the shuffle mask
+  const uint8_t id = patterns_id[hashcode];
+  if (id >= 81) {
+    return 0;
+  }
+  const uint8_t *pat = &patterns[id][0];
+  const __m128i pattern = _mm_loadu_si128((const __m128i *)pat);
+  // The value of the shuffle mask at a specific index points at the last digit,
+  // we check that it matches the length of the input.
+  const __m128i ascii0 = _mm_set1_epi8('0');
+  const __m128i t0 = input;
+  __m128i t1 = _mm_shuffle_epi8(t0, pattern);
+  // check that leading digits of 2- 3- numbers are not zeros.
+  {
+    const __m128i eq0 = _mm_cmpeq_epi8(t1, ascii0);
+    if (!_mm_testz_si128(eq0, _mm_set_epi8(-1, 0, -1, 0, -1, 0, -1, 0,
+                                           0, 0, 0, 0, 0, 0, 0, 0))) {
+      return 0;
+    }
+  }
+  // replace null values with '0'
+  __m128i t1b = _mm_blendv_epi8(t1, ascii0, pattern);
+
+  // subtract '0'
+  const __m128i t2 = _mm_sub_epi8(t1b, ascii0);
+  // check that everything was in the range '0' to '9'
+  {
+    const __m128i c9 = _mm_set1_epi8('9' - '0');
+    const __m128i t2m = _mm_max_epu8(t2, c9);
+    const __m128i t2me = _mm_cmpeq_epi8(t2m, c9);
+    if (!_mm_test_all_ones(t2me)) {
+      return 0;
+    }
+  }
+  // We do the computation, the Mula way.
+  const __m128i weights =
+      _mm_setr_epi8(1, 10, 1, 10, 1, 10, 1, 10, 100, 0, 100, 0, 100, 0, 100, 0);
+  const __m128i t3 = _mm_maddubs_epi16(t2, weights);
+  const __m128i t4 = _mm_alignr_epi8(t3, t3, 8);
+  const __m128i t5 = _mm_add_epi16(t4, t3);
+  // Test that we don't overflow (over 255)
+  if (!_mm_testz_si128(t5, _mm_set_epi8(0, 0, 0, 0, 0, 0, 0, 0, -1, 0, -1,
+                                        0, -1, 0, -1, 0))) {
+    return 0;
+  }
+  // pack and we are done!
+  const __m128i t6 = _mm_packus_epi16(t5, t5);
+  uint32_t address =  (uint32_t)_mm_cvtsi128_si32(t6);
+  memcpy(destination, &address, 4);
+  return (int)(ipv4_string_length - (size_t)pat[6]);
```

Comment:
needs coercion to bool
`return (int)((ipv4_string_length - (size_t)pat[6]) == 1);`

Reply:
The function returns 1 on success. It returns an integer value different from 1 on failure.

Reply:
Please check how it is used:

```
if (sse_inet_aton_16(token->data, &parser->rdata->octets[parser->rdata->length]) != 1)
```

(This is basically how @k0ekk0ek wrote it.)